src/main/java/io/cdap/plugin/gcp/publisher/source/PubSubReceiver.java

```diff
@@ -0,0 +1,506 @@
+        AcknowledgeRequest.newBuilder()
+          .setSubscription(subscription)
+          .addAllAckIds(ackIds)
+          .build();
+      subscriber.acknowledgeCallable().call(acknowledgeRequest);
+    }
+  }
+
+  /**
+   * Get Subscriber settings instance.
+   *
+   * @return the Subscriber Stub settings needed to subscribe to a Pub/Sub topic.
+   */
+  protected SubscriberStubSettings getSubscriberSettings() throws IOException {
+    SubscriberStubSettings.Builder builder = SubscriberStubSettings.newBuilder();
+
+    if (credentials != null) {
+      builder.setCredentialsProvider(FixedCredentialsProvider.create(credentials));
+    }
+
+    return builder.build();
+  }
+
+  /**
+   * Get Subscription Admin Client instance
+   *
+   * @return Subscription Admin Client instance.
+   * @throws IOException
+   */
+  protected SubscriptionAdminClient getSubscriptionAdminClient() throws IOException {
+    return SubscriptionAdminClient.create();
+  }
+
+  /**
+   * Get Subscriber Stub instance
+   *
+   * @param subscriberStubSettings The settings for this Subscriber Stub client.
+   * @return a new Subscriber Stub Instance
+   * @throws IOException the exception thrown by the Pub/Sub library if the Subscriber Stub Settings are invalid.
+   */
+  protected SubscriberStub getSubscriber(SubscriberStubSettings subscriberStubSettings) throws IOException {
+    return GrpcSubscriberStub.create(subscriberStubSettings);
+  }
+
+  /**
+   * Sleep for a given number of milliseconds, calculate new backoff time and return.
+   *
+   * @param backoff the time in milliseconds to delay execution.
+   * @return the new backoff delay in milliseconds
+   */
+  protected int sleepAndIncreaseBackoff(int backoff) {
+    try {
+      if (!isStopped()) {
+        LOG.debug("Backoff - Sleeping for {} ms.", backoff);
+        Thread.sleep(backoff);
+      }
+    } catch (InterruptedException e) {
+      LOG.warn("Interrupted Exception in PubSubReceiver.");
+    }
+
+    return calculateUpdatedBackoff(backoff);
+  }
+
+  /**
+   * Calculate the updated backoff period baded on the Backoff configuration parameters.
+   *
+   * @param backoff the previous backoff period
+   * @return the updated backoff period for the next cycle.
+   */
+  protected int calculateUpdatedBackoff(int backoff) {
+    return Math.min((int) (backoff * backoffConfig.getBackoffFactor()), backoffConfig.getMaximumBackoffMs());
+  }
+
+  /**
+   * Get the rate at which this receiver should pull messages and set this rate in the bucket we use for rate control.
+   * The default rate is Integer.MAX_VALUE if the receiver has not been able to calculate a rate.
+   */
+  protected void updateMessageRateAndFillBucket() {
+    int messageRate = (int) Math.min(supervisor().getCurrentRateLimit(), Integer.MAX_VALUE);
+
+    if (messageRate != previousFetchRate) {
+      previousFetchRate = messageRate;
+      LOG.debug("Receiver fetch rate is set to: {}", messageRate);
```

Comment:
better to limit how often this logs, as debug logs are still collected by default. Maybe max once a minute?

Reply:
@albertshau What do you think about using `trace` as the log level?

Reply:
that's fine too